AwsCryptographicMaterialProviders/runtimes/java/src/testExamples/java/software/amazon/cryptography/example/hierarchy/concurrent/StorageVersionReadConcurrencyTests.java

```diff
@@ -0,0 +1,302 @@
+  public void testConcurrentVersionWithStorage() {
+    String threadId = String.valueOf(Thread.currentThread().getId());
+    String threadIdToIndex = storageIndexToThreadId.computeIfAbsent(
+      threadId,
+      str -> unpickedIndicesForStorage.pop()
+    );
+
+    String timestamp = dateFormat.format(new Date());
+    try {
+      Storage threadStorage = storageForThread(threadIdToIndex);
+      KeyStoreAdminConfig keyStoreAdminConfig = KeyStoreAdminConfig
+        .builder()
+        .storage(threadStorage)
+        .logicalKeyStoreName(Fixtures.TEST_KEYSTORE_NAME)
+        .build();
+      KeyStoreAdmin admin = KeyStoreAdmin
+        .builder()
+        .KeyStoreAdminConfig(keyStoreAdminConfig)
+        .build();
+      raceToVersionWithStorage(admin);
+      System.out.println(
+        "Successfully versioned branch key! Thread ID: " +
+        Thread.currentThread().getId() +
+        " ThreadIndex: " +
+        threadIdToIndex +
+        " Timestamp: " +
+        timestamp +
+        " BranchKeyId: " +
+        branchKeyId
+      );
+    } catch (TransactionCanceledException exception) {
+      System.out.println("Failed to write branch key: " + branchKeyId);
+      // Exceptions that get thrown when you write keys using the Storage interface
+      exception
+        .cancellationReasons()
+        .forEach(cancellationReason -> {
+          Assert.assertTrue(
+            (cancellationReason.code().equals("TransactionConflict") ||
+              cancellationReason.code().equals("None") ||
+              cancellationReason.code().equals("ConditionalCheckFailed"))
+          );
+        });
+    }
+  }
+
+  @Test(threadPoolSize = 10, invocationCount = 100, timeOut = 10000)
+  public void testConcurrentActiveReadWhileVersionInFlight() {
+    // Since on set up we create a branch key, we should always be able to read.
+    String threadId = String.valueOf(Thread.currentThread().getId());
+    String threadIdToIndex = keyStoreIndexToThreadId.computeIfAbsent(
+      threadId,
+      str -> unpickedIndicesForKeyStore.pop()
+    );
+    KeyStore keyStore = keyStoreForThread(threadIdToIndex);
+    GetActiveBranchKeyOutput output = raceToReadActiveWithKeyStore(keyStore);
+    versionKeyOutputMap.put(
+      output.branchKeyMaterials().branchKeyVersion(),
+      output.branchKeyMaterials().branchKey()
+    );
+    System.out.println(
+      "Successfully read branch key: " +
+      branchKeyId +
+      " with version: " +
+      output.branchKeyMaterials().branchKeyVersion()
+    );
+  }
+
+  @Test(
+    dependsOnMethods = {
+      "testConcurrentActiveReadWhileVersionInFlight",
+      "testConcurrentVersionWithStorage",
+    }
+  )
+  public void testVersionReads() {
+    Assert.assertFalse(versionKeyOutputMap.isEmpty());
+    for (String key : versionKeyOutputMap.keySet()) {
+      System.out.println(
+        "key: " + key + " value " + versionKeyOutputMap.get(key)
+      );
+    }
+  }
```

Comment:
This method asserts that `testConcurrentActiveReadWhileVersionInFlight` succeeded at least once.

Reply:
Yes. That is what I want. This test proves that while a versioning is going on we can still read.